worldedit-core/src/main/java/com/boydti/fawe/beta/implementation/lighting/NMSRelighter.java

```diff
@@ -0,0 +1,637 @@
+        if (lightLock.compareAndSet(false, true)) {
+            synchronized (lightQueue) {
+                try {
+                    long[][][] currentMap = lightQueue.get(index);
+                    if (currentMap == null) {
+                        currentMap = new long[16][][];
+                        this.lightQueue.put(index, currentMap);
+                    }
+                    set(x & 15, y, z & 15, currentMap);
+                } finally {
+                    lightLock.set(false);
+                }
+            }
+        } else {
+            long[][][] currentMap = concurrentLightQueue.get(index);
+            if (currentMap == null) {
+                currentMap = new long[16][][];
+                this.concurrentLightQueue.put(index, currentMap);
+            }
+            set(x & 15, y, z & 15, currentMap);
+        }
+    }
+
+    public synchronized void clear() {
+        extentdSkyToRelight.clear();
+        skyToRelight.clear();
+        chunksToSend.clear();
+        lightQueue.clear();
+    }
+
+    public boolean addChunk(int cx, int cz, byte[] fix, int bitmask) {
+        RelightSkyEntry toPut = new RelightSkyEntry(cx, cz, fix, bitmask);
+        extentdSkyToRelight.add(toPut);
+        return true;
+    }
+
+    private synchronized Map<Long, RelightSkyEntry> getSkyMap() {
+        RelightSkyEntry entry;
+        while ((entry = extentdSkyToRelight.poll()) != null) {
+            long pair = MathMan.pairInt(entry.x, entry.z);
+            RelightSkyEntry existing = skyToRelight.put(pair, entry);
+            if (existing != null) {
+                entry.bitmask |= existing.bitmask;
+                if (entry.fix != null) {
+                    for (int i = 0; i < entry.fix.length; i++) {
+                        entry.fix[i] &= existing.fix[i];
+                    }
+                }
+            }
+        }
+        return skyToRelight;
+    }
+
+    public synchronized void removeLighting() {
+        Iterator<Map.Entry<Long, RelightSkyEntry>> iter = getSkyMap().entrySet().iterator();
+        while (iter.hasNext()) {
+            Map.Entry<Long, RelightSkyEntry> entry = iter.next();
+            RelightSkyEntry chunk = entry.getValue();
+            long pair = entry.getKey();
+            Integer existing = chunksToSend.get(pair);
+            chunksToSend.put(pair, chunk.bitmask | (existing != null ? existing : 0));
+            ChunkHolder iChunk = (ChunkHolder) queue.getOrCreateChunk(chunk.x, chunk.z);
+            if (!iChunk.isInit()) {
+                iChunk.init(queue, chunk.x, chunk.z);
+            }
+            for (int i = 0; i < 16; i++) {
+                iChunk.removeSectionLighting(i, true);
+            }
+            iter.remove();
+        }
+    }
+
+    public void updateBlockLight(Map<Long, long[][][]> map) {
+        int size = map.size();
+        if (size == 0) {
+            return;
+        }
+        Queue<MutableBlockVector3> lightPropagationQueue = new ArrayDeque<>();
```

Comment:
Consider allocating these with a reasonable initial capacity to prevent it from being expanded when new items are added because speed and stuff